src/uopool/services/sanity_check.rs

```diff
@@ -0,0 +1,559 @@
+                    )
+                    .is_ok()
+                {
+                    self.sanity_check_results
+                        .write()
+                        .entry(user_operation.hash(&entry_point.address(), &self.chain_id))
+                        .or_insert_with(Default::default)
+                        .insert(SanityCheckResult::SenderVerified);
+                    return Ok(());
+                }
+            }
+        }
+
+        if let Some(mempool) = self.mempools.write().get(&mempool_id) {
+            let user_operation_prev = mempool.get_all_by_sender(&user_operation.sender)[0].clone();
+            let fee_per_gas_diff = user_operation.max_priority_fee_per_gas
+                - user_operation_prev.max_priority_fee_per_gas;
+
+            if user_operation.sender == user_operation_prev.sender
+                && user_operation.nonce == user_operation_prev.nonce
+                && user_operation.max_priority_fee_per_gas
+                    > user_operation_prev.max_priority_fee_per_gas
+                && (user_operation.max_fee_per_gas - user_operation_prev.max_fee_per_gas)
+                    == fee_per_gas_diff
+            {
+                return Ok(());
+            }
+        }
+
+        Err(BadUserOperationError::SenderVerification {
+            sender: user_operation.sender,
+        })
+    }
+
+    pub async fn validate_user_operation(
+        &self,
+        user_operation: &UserOperation,
+        entry_point: &Address,
+    ) -> Result<(), BadUserOperationError<M>> {
+        self.sanity_check_results.write().insert(
+            user_operation.hash(entry_point, &self.chain_id),
+            Default::default(),
+        );
+
+        // Either the sender is an existing contract, or the initCode is not empty (but not both)
+        self.sender_or_init_code(user_operation).await?;
+
+        // If initCode is not empty, parse its first 20 bytes as a factory address. Record whether the factory is staked, in case the later simulation indicates that it needs to be. If the factory accesses global state, it must be staked - see reputation, throttling and banning section for details.
+        self.verify_factory(user_operation, entry_point).await?;
+
+        // The verificationGasLimit is sufficiently low (<= MAX_VERIFICATION_GAS) and the preVerificationGas is sufficiently high (enough to pay for the calldata gas cost of serializing the UserOperation plus PRE_VERIFICATION_OVERHEAD_GAS)
+        self.verification_gas(user_operation)?;
+
+        // The paymasterAndData is either empty, or start with the paymaster address, which is a contract that (i) currently has nonempty code on chain, (ii) has a sufficient deposit to pay for the UserOperation, and (iii) is not currently banned. During simulation, the paymaster's stake is also checked, depending on its storage usage - see reputation, throttling and banning section for details.
+        self.verify_paymaster(user_operation, entry_point).await?;
```

Comment:
It seems like `verify_paymaster` doesn't check the `(i)` currently.

Reply:
Ah, missed that. Tnx